updated version with new features

zingale opened the pull request.

xhf-F commented:
I am just confused that when you add cross product, it seems that you didn't add a testing assert so when you call pull request the cpp file didn't test this cross product therefore you pass the checks. Right? (Btw, I know you did these just for changing back to the right version and adding something for a new version to call pull request.)

zingale commented:
you are right -- I did two things at once -- I fixed the bug we introduced and I added the cross product.

even though I did not add a test for the cross product, it still ran all of the other tests.  In reality, it is always a good idea to add tests at the same time you add new features, so I should have done so.
